plugin/ochttp/trace.go

```diff
@@ -157,31 +157,48 @@ type Handler struct {
 	Handler http.Handler
 }
 
+type ender interface {
+	End()
+}
+
 func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var traceEnd, statsEnd ender
 	if !h.NoTrace {
-		name := spanNameFromURL("Recv", r.URL)
-		p := h.Propagation
-		if p == nil {
-			p = defaultFormat
-		}
-		ctx := r.Context()
-		var span *trace.Span
-		if sc, ok := p.SpanContextFromRequest(r); ok {
-			ctx, span = trace.StartSpanWithRemoteParent(ctx, name, sc, trace.StartOptions{})
-		} else {
-			ctx, span = trace.StartSpan(ctx, name)
-		}
-		defer span.End()
-
-		span.SetAttributes(requestAttrs(r)...)
-		r = r.WithContext(ctx)
+		w, r, traceEnd = h.wrapForTrace(w, r)
+	}
+	if !h.NoStats {
+		w, r, statsEnd = h.wrapForStats(w, r)
 	}
 
 	handler := h.Handler
 	if handler == nil {
 		handler = http.DefaultServeMux
 	}
 	handler.ServeHTTP(w, r)
+	if traceEnd != nil {
+		traceEnd.End()
+	}
+	if statsEnd != nil {
+		statsEnd.End()
+	}
+}
+
+func (h *Handler) wrapForTrace(w http.ResponseWriter, r *http.Request) (http.ResponseWriter, *http.Request, ender) {
```

Comment:
Perhaps registerReceiverSpan?

I couldn't make out what wrapForTrace intended to convey.

Reply:
The incoming Request and ResponseWriter are being wrapped in a new context and the tracking writer. eh, naming is hard.

I'll go with startTrace per @rakyll